internal/snapshots/snapshotter.go

```diff
@@ -0,0 +1,444 @@
+					- #len(lineHashes) number of fseeks per copy block -> not good, saves memory but huge IO overhead
+					- Preferably, sort based on the first column or a unique column for data locality
+				*/
+
+				// Write the sorted data to the output file based on the offset
+				for _, hashedLine := range lineHashes {
+					// Seek to the offset of the line in the input file
+					_, err := dumpInst2.Seek(hashedLine.offset, 0)
+					if err != nil {
+						return nil, fmt.Errorf("failed to seek to offset: %w", err)
+					}
+
+					lineBytes, err := bufio.NewReader(dumpInst2).ReadBytes('\n')
+					if err != nil {
+						return nil, fmt.Errorf("failed to read line from input file: %w", err)
+					}
+
+					_, err = outputFile.Write(lineBytes)
+					if err != nil {
+						return nil, fmt.Errorf("failed to write to sanitized dump file: %w", err)
+					}
+				}
+
+				// Write the end of COPY block to the output file
+				_, err = outputFile.WriteString(line + "\n")
+				if err != nil {
+					return nil, fmt.Errorf("failed to write to sanitized dump file: %w", err)
+				}
+
+				// Clear the lineHashes array
+				lineHashes = make([]HashedLine, 0)
+				offset += numBytes // +1 for newline character
+			} else {
+				// If we are in a COPY block, we need to sort the data based on the row hash
+				hasher := sha256.New()
+				hasher.Write([]byte(line))
+				lineHashes = append(lineHashes, HashedLine{Hash: hasher.Sum(nil), offset: offset})
+				offset += numBytes // +1 for newline character
+			}
+		} else {
+			offset += int64(len(line)) + 1 // +1 for newline character
+			if line == "" {
+				// skip empty lines
+				continue
+			} else if strings.HasPrefix(line, "--") {
+				// skip comments
+				continue
+			} else if strings.HasPrefix(line, "SET") || strings.HasPrefix(line, "SELECT") {
+				// skip SET and SELECT statements
+				continue
+			} else {
+				if strings.HasPrefix(line, "COPY") && strings.Contains(line, "FROM stdin;") {
+					// start of COPY block
+					inCopyBlock = true
+				}
+				// write the line to the output file
+				_, err := outputFile.WriteString(line + "\n")
+				if err != nil {
+					return nil, fmt.Errorf("failed to write to sanitized dump file: %w", err)
+				}
+			}
+		}
+	}
+
+	if err := scanner.Err(); err != nil {
+		return nil, fmt.Errorf("failed to scan the dump file: %w", err)
+	}
+
+	// remove the dump file
+	err = os.Remove(dumpFile)
+	if err != nil {
+		return nil, fmt.Errorf("failed to remove dump file: %w", err)
+	}
+
+	hash, err := utils.HashFile(sanitizedDumpFile)
+	if err != nil {
+		return nil, fmt.Errorf("failed to hash sanitized dump file: %w", err)
+	}
+
+	s.log.Info("Sanitized dump file", zap.Uint64("height", height), zap.String("dumpFile hash", fmt.Sprintf("%x", hash)))
+
+	return hash, nil
+}
+
+// CompressDump is the STAGE3 of the snapshot creation process
+// This method compresses the sanitized dump file using gzip compression
+// Should we do inline compression? or using exec.Command?
```

Comment:
Inline seems right here. I see no reason to use exec, it just introduces more room for issues